File: internal/app/triggers.go
package app

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
)

// TriggerWebhook sends a POST to the device's webhook URL
func TriggerWebhook(d *Device, colors Colors) {
	body, err := json.Marshal(map[string]interface{}{
		"colors": colors.AsBooleans(),
	})
	if err != nil {
		log.Printf("TriggerWebhook marshal error: %v", err)
		return
	}

	req, err := http.NewRequest("POST", d.WebhookURL, bytes.NewReader(body))
	if err != nil {
		log.Printf("TriggerWebhook request error: %v", err)
		return
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("x-ryg", colors.RYG())
	req.Header.Set("x-device-url", fmt.Sprintf("https://%s/api/devices/%s", AppHost, d.ID))

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		log.Printf("TriggerWebhook error for device %s: %v", d.Name, err)
		return
	}
	resp.Body.Close()
}

// TriggerParticle publishes a build_state event to Particle
func TriggerParticle(d *Device) {
	if ParticleAccessToken == "" {
		return
	}

	body, _ := json.Marshal(map[string]interface{}{
		"name":    "build_state",
		"data":    d.Status,
		"ttl":     3600,
		"private": false,
	})

	req, err := http.NewRequest("POST", "https://api.particle.io/v1/devices/events", bytes.NewReader(body))
	if err != nil {
		log.Printf("TriggerParticle request error: %v", err)
		return
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+ParticleAccessToken)

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		log.Printf("TriggerParticle error: %v", err)
		return
	}
	resp.Body.Close()
}

File: src/triggers.zig
const std = @import("std");
const models = @import("models.zig");

/// Send a POST to the device's webhook URL with colors as JSON body.
/// Fire-and-forget: errors are logged but do not propagate.
pub fn triggerWebhook(
    allocator: std.mem.Allocator,
    device: models.Device,
    colors: models.Colors,
    host: []const u8,
) void {
    // Build JSON body
    var body_buf: [256]u8 = undefined;
    var fbs = std.io.fixedBufferStream(&body_buf);
    const writer = fbs.writer();
    writer.writeAll("{\"colors\":") catch return;
    colors.writeJsonBooleans(writer) catch return;
    writer.writeByte('}') catch return;
    const body = fbs.getWritten();

    // Build RYG header value
    var ryg_buf: [3]u8 = undefined;
    const ryg = colors.ryg(&ryg_buf);

    // Build device URL header value
    var url_buf: [256]u8 = undefined;
    const device_url = std.fmt.bufPrint(&url_buf, "https://{s}/api/devices/{s}", .{ host, device.id }) catch return;

    // Make HTTP request using fetch
    var client: std.http.Client = .{ .allocator = allocator };
    defer client.deinit();

    const uri = std.Uri.parse(device.webhook_url) catch {
        std.log.err("TriggerWebhook: invalid URL {s}", .{device.webhook_url});
        return;
    };

    _ = client.fetch(.{
        .location = .{ .uri = uri },
        .method = .POST,
        .payload = body,
        .extra_headers = &.{
            .{ .name = "content-type", .value = "application/json" },
            .{ .name = "x-ryg", .value = ryg },
            .{ .name = "x-device-url", .value = device_url },
        },
    }) catch |err| {
        std.log.err("TriggerWebhook: fetch error {}", .{err});
        return;
    };
}

/// Publish a build_state event to Particle.io.
/// Fire-and-forget: errors are logged but do not propagate.
pub fn triggerParticle(
    allocator: std.mem.Allocator,
    status: []const u8,
    token: ?[]const u8,
) void {
    const access_token = token orelse return;
    if (access_token.len == 0) return;

    // Build JSON body
    var body_buf: [256]u8 = undefined;
    const body = std.fmt.bufPrint(&body_buf,
        \\{{"name":"build_state","data":"{s}","ttl":3600,"private":false}}
    , .{status}) catch return;

    // Build auth header value
    var auth_buf: [256]u8 = undefined;
    const auth = std.fmt.bufPrint(&auth_buf, "Bearer {s}", .{access_token}) catch return;

    var client: std.http.Client = .{ .allocator = allocator };
    defer client.deinit();

    _ = client.fetch(.{
        .location = .{ .url = "https://api.particle.io/v1/devices/events" },
        .method = .POST,
        .payload = body,
        .extra_headers = &.{
            .{ .name = "content-type", .value = "application/json" },
            .{ .name = "authorization", .value = auth },
        },
    }) catch |err| {
        std.log.err("TriggerParticle: fetch error {}", .{err});
        return;
    };
}
